### PR TITLE
Implement and use CurrentProductContext & useCurrentProduct hook

### PR DESCRIPTION
**Pull request by ashleemboyer**
## "Ready for Review" checklist

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## Relevant links

- [Preview link](https://dev-portal-git-ambcurrent-product-context-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1201010428539925/1201602267333006/f) 🎟️

### Main additions

- Added `CurrentProductProvider` and renders/hydrates it in `src/pages/_app.tsx`
- Added `useCurrentProduct` hook for consuming the value of `CurrentProductContext`
- Added a test suite for `CurrentContextProvider` (`src/contexts/__tests__/current-product.test.tsx`)

### Supplemental additions

- Added JSON files in `src/data` for Boundary, Consul, HashiCorp Cloud Platform, Nomad, Packer, Sentinel, Terraform, Vagrant, and Vault
- Added placeholder home pages for Boundary, Consul, HashiCorp Cloud Platform, Nomad, Packer, Sentinel, Terraform, and Vagrant
- Added `NavigationHeaderItem` interface for use in the `Product` interface, among other useful properties

### Refactors

- Updated `config/products.ts` to "generate" product urls from their slugs (just prepends `/` to the slug) now that we have internal placeholder pages to point to
- Updated `src/components/docs-anchor/index.tsx` to just be a component (no makeDocsAnchor) and leverage `useCurrentProduct`, and updated the tests
- Updated `DocsPage` to leverage `useCurrentProduct` and memoize `additionalComponents`
- Updated `NavigationHeader` to leverage `useCurrentProduct`
- Updated `ProductSwitcher` to leverage `useCurrentProduct` (and fixed the Vault icon when it is the current product being viewed)
- Moved Vault's nav data from `src/data/vault-landing.json` to `src/data/vault.json`
- Moved Waypoint's nav data from `src/data/waypoint-landing.json` to `src/data/waypoint.json`
- Updated `src/layouts/docs/server.ts` and `src/views/product-landing/server.ts` to build navData from `data` files and additionally return `product` in `finalProps`
- Updated `src/pages/vault/index.tsx` to read Vault's data from `data/vault.json`
- Updated Waypoint's commands, docs, and plugins pages to have unique component names and read Waypoint's data from `data/waypoint.json`
- Updated `src/pages/waypoint/downloads/index.tsx` to read Waypoint's data from `data/waypoint.json`

### Other

- Removed `src/pages/waypoint/community/index.tsx` and `src/pages/waypoint/community/style.module.css` because that page is no longer a thing in DevDot
- Removed `config/waypoint.json` and `src/config/waypoint.json` and moved their content to `data/waypoint.json` to consolidate where we source data from
- Added some configuration to `jest.config.ts` to fix an error: `Cannot find module 'contexts/current-product' from 'src/contexts/__tests__/current-product.test.tsx'`
- Removed unused imports from `src/components/product-icon/index.tsx` and `src/layouts/base-new/index.tsx`

## How

Not much to add here since most of these changes were just moving around already existing code. I did find [this React Hooks documentation](https://react-hooks-testing-library.com/usage/advanced-hooks#context) helpful for creating the `setup` function in the `CurrentProductContext` test suite.

## Testing

- [ ] Go to https://dev-portal-git-ambcurrent-product-context-hashicorp.vercel.app/
  - [ ] The ProductSwitcher should say "Products" and the page content should be correct
- [ ] For each of the products listed in ProductSwitcher, complete these steps:
  - [ ] Open the ProductSwitcher
  - [ ] Select a product
  - [ ] The ProductSwitcher should have the correct product name and icon
  - [ ] The page content should be correct
- [ ] Click the HashiCorp logo at the top left of the page to go to the main home page
  - [ ] The ProductSwitcher should update to say `Products`
- [ ] Go to https://dev-portal-git-ambcurrent-product-context-hashicorp.vercel.app/vault
  - [ ] The sidebar links should still be correct (all are currently external)
- [ ] Go to https://dev-portal-git-ambcurrent-product-context-hashicorp.vercel.app/waypoint
  - [ ] The sidebar links should all still be correct
  - [ ] The Overview, Docs, CLI, Plugins, and Install links should:
    - [ ] still link to the correct pages
    - [ ] have the pages working as expected
    - [ ] show "Waypoint" in the ProductSwitcher
- [ ] Go to https://dev-portal-git-ambcurrent-product-context-hashicorp.vercel.app/
  - [ ] Click on a product card in the page body
  - [ ] You should be taken to the correct page
  - [ ] The ProductSwitcher should show the correct product name and icon
- [ ] Click the HashiCorp logo at the top left to go back to the home page
  - [ ] The ProductSwitcher should show "Products""
  - [ ] Click the back button
  - [ ] The ProductSwitcher should show the correct name and icon of the product you were viewing before clicking the logo

## Anything else?

- Yellow Vault icon coming soon: https://github.com/hashicorp/flight/issues/372

The following was fixed by not using `useState` in `CurrentContextProvider` and just setting the value of the context based on whether or not the current router path is `/` or not:

- There's a bug I'm not sure how to handle yet. I think we have a few options but would like to gather ideas. Steps to reproduce:
  - Go to https://dev-portal-git-ambcurrent-product-context-hashicorp.vercel.app/
  - Select a product from the ProductSwitcher
  - Click the HashiCorp icon at the top left to go to the main home page
  - Click the back button
  - The ProductSwitcher still says "Products" instead of whatever product you were previously looking at

### CODE REVIEW

**Comment by ashleemboyer** [comment]
Requesting some early reviews before I add tests and docs 🙏🏻

**Comment by BRKalow** [comment]
@ashleemboyer do you think it'd be valuable to have any additional navigation related data in a context like this? Thinking back to my initial message:

```ts
interface HashiNavigationContext {
  product
  productBasePaths
  acitveBasePath
  hash
}
```

(totally fine if no, I'm guessing you gave this some thought during implementation!)

File: src/contexts/current-product.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  children,
+  currentProduct,
+}) => {
+  const router = useRouter()
+  const [value, setValue] = useState(currentProduct)
+
+  /**
+   * This effect is for clearing the current product on pages that do not have a
+   * single product. This currently includes the main home page.
+   */
```
Comment (BRKalow) [inline src/contexts/current-product.tsx:17]: 💭  👍

File: src/contexts/current-product.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  )
+}
+
+const useCurrentProduct = (): Product => {
+  const context = useContext(CurrentProductContext)
+  if (context === undefined) {
+    throw new Error(
+      'useCurrentProduct must be used within a CurrentProductProvider'
+    )
+  }
```
Comment (BRKalow) [inline src/contexts/current-product.tsx:43]: Great error message!

File: src/components/docs-anchor/index.tsx [hunk 3]
```diff
@@ -1,47 +1,46 @@
...
-          <a {...passableProps}>{children}</a>
-        </Link>
-      )
-    }
+  let adjustedHref = href
+  // TODO: infer this condition dynamically
+  if (
+    currentProduct.basePaths.some((basePath) => href.startsWith(`/${basePath}`))
+  ) {
+    // TODO: infer the product dynamically
```
Comment (BRKalow) [inline src/components/docs-anchor/index.tsx:25]: I think this TODO is done now!

```suggestion
```

**Review by BRKalow — COMMENTED** [review]
Looking really good 😃

File: src/pages/index.tsx [hunk 4]
```diff
@@ -5,7 +5,7 @@ import HomepageHero from 'components/homepage-hero'
 import Footer from 'components/footer'
 import s from './index.module.css'
 
-function Homepage(): ReactElement {
+const WaypointHomePage = (): ReactElement => {
```
Comment (zchsh) [inline src/pages/index.tsx:8]: I think this might have been an accidental rename, since I think this is the main `/` homepage!

File: src/layouts/docs/server.ts [hunk 5]
```diff
@@ -4,6 +4,7 @@ import {
 } from '@hashicorp/react-docs-page/server'
 import { anchorLinks } from '@hashicorp/remark-plugins'
 
+import { Product } from 'types/products'
```
Comment (zchsh) [inline src/layouts/docs/server.ts:7]: 👍 Nice, was just thinking we should add this type!

File: src/types/products.ts [hunk 6]
```diff
@@ -1,10 +1,21 @@
...
 interface Product extends ProductMeta {
+  basePaths?: string[]
+  navigationHeaderItems?: NavigationHeaderItem[]
+  sidebar?: {
+    landingPageNavData: MenuItem[]
+    resourcesNavData: MenuItem[]
+  }
   slug: ProductSlug
-  url: string
+  url?: string
```
Comment (zchsh) [inline src/types/products.ts:18]: I think we also need a `name` property here - used in the waypoint product landing `server.ts`, and also in the docs layout `server.ts`. Alternately I guess we could always derive the product name from the slug? Though I worry about edge cases with eg. HCP or sub-products, where we might want to pass a specific product slug but have more flexibility in the displayed name (eg `HCP` / `HashiCorp Cloud Platform`).

Comment (ashleemboyer) [reply src/types/products.ts:18]: Ah, yes! `name` is very necessary just slipped the mind to add it so far. 😊

Names are actually currently derived here: https://github.com/hashicorp/dev-portal/blob/assembly-ui-v1/config/products.ts#L40. I'm pretty sure I referenced some existing code for that but I don't remember where it is.

File: src/pages/index.tsx (hunk 4 above)
Comment (ashleemboyer) [reply src/pages/index.tsx:8]: Oops, it was! I was renaming a few others that all had the same name and I think I accidentally grouped this one in with it too. Will fix. 😊

File: src/components/docs-anchor/index.tsx (hunk 3 above)
Comment (ashleemboyer) [reply src/components/docs-anchor/index.tsx]: Great catch! Updated ✅

**Comment by ashleemboyer** [comment]
> @ashleemboyer do you think it'd be valuable to have any additional navigation related data in a context like this? Thinking back to my initial message:
> 
> ```ts
> interface HashiNavigationContext {
>   product
>   productBasePaths
>   acitveBasePath
>   hash
> }
> ```
> 
> (totally fine if no, I'm guessing you gave this some thought during implementation!)

@BRKalow `basePaths` will get included in the Context I've added - see [`src/data/waypoint.json`](https://github.com/hashicorp/dev-portal/blob/amb.current-product-context/src/data/waypoint.json#L4). My idea right now is to make sure that's configured in every product's file in `src/data` and that's something I'll make sure to document. It's in the TS interface defined for a Product, but I'm thinking I'll create another interface so that it's more clear where that property is sourced from. I already have a TODO in place for that where I was wondering if it was necessary, I think it is now.

`activeBasePath` sounds like a great addition! That has to be derived from the route and making the Context do that is way easier than doing it manually every time. Might put that in the route change listener I made for clearing out the Context value on `/`.

Have you identified a use for `hash`? Nothing is coming to mind at the moment but that doesn't mean I think we should exclude it. Just trying to get a bigger picture of the proposal. 😊

**Comment by BRKalow** [comment]
No concrete use for `hash` at the moment! Sounds like you've got everything covered 🗺️

**Comment by ashleemboyer** [comment]
Note to self, something to implement after Zach's PR #69 is merged: https://github.com/hashicorp/dev-portal/pull/69#discussion_r781644656

File: src/components/docs-page/index.tsx [hunk 7]
```diff
@@ -1,27 +1,27 @@
...
+  const currentProduct = useCurrentProduct()
+
+  const additionalComponents = defaultMdxComponents({
+    product: currentProduct.slug,
+    additionalComponents: {
+      Placement,
+      NestedNode,
+      a: DocsAnchor,
+    },
+  })
```
Comment (BRKalow) [inline src/components/docs-page/index.tsx:23]: This should be memoized based on `currentProduct.slug`, to ensure that the component references don't get re-created on re-render, potentially causing full remounts.

```suggestion
  const additionalComponents = useMemo(() => 
    defaultMdxComponents({
	    product: currentProduct.slug,
	    additionalComponents: {
	      Placement,
	      NestedNode,
	      a: DocsAnchor,
	    },
	  }), [currentProduct.slug])
```

File: src/contexts/__tests__/current-product.test.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+const setup = (currentProduct: Product) => {
+  const wrapper = ({ children }) => (
+    <CurrentProductProvider currentProduct={currentProduct}>
+      {children}
+    </CurrentProductProvider>
+  )
+  return renderHook(() => useCurrentProduct(), { wrapper })
+}
+
+describe('CurrentProductContext', () => {
```
Comment (BRKalow) [inline src/contexts/__tests__/current-product.test.tsx:24]: nice tests!

**Review by BRKalow — COMMENTED** [review]
The code looks good to me overall!

Noticed in preview:
* Product switcher is not updating when navigating to a new product via the homepage cards

File: src/components/docs-page/index.tsx (hunk 7 above)
Comment (ashleemboyer) [reply src/components/docs-page/index.tsx]: Made this change! ✅

File: src/contexts/__tests__/current-product.test.tsx (hunk 8 above)
Comment (ashleemboyer) [reply src/contexts/__tests__/current-product.test.tsx:23]: Thank you! 😊

File: src/contexts/__tests__/current-product.test.tsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
```
Comment (BRKalow) [inline src/contexts/__tests__/current-product.test.tsx:23]: With the refactor away from effects, do we need a test case to validate that it sets the context value to `null` when the path is `/`?

File: src/contexts/current-product.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+type CurrentProduct = Product | undefined
+
+const CurrentProductContext = createContext<CurrentProduct>(undefined)
+
+const CurrentProductProvider: FC<{ currentProduct: Product }> = ({
+  children,
+  currentProduct,
+}) => {
+  const router = useRouter()
+  const value = router.asPath === '/' ? null : currentProduct
```
Comment (BRKalow) [inline src/contexts/current-product.tsx:16]: clean

**Review by BRKalow — APPROVED** [review]
Looks great! One comment about a potential test case, but non-blocking.

File: src/contexts/__tests__/current-product.test.tsx (hunk 9 above)
Comment (ashleemboyer) [reply src/contexts/__tests__/current-product.test.tsx:23]: Good idea, I'll give that a try now 😊

Comment (ashleemboyer) [reply src/contexts/__tests__/current-product.test.tsx:23]: Added! What do you think? See https://github.com/hashicorp/dev-portal/pull/67/commits/96981029061087bb94433cd4587a727db843dff2

Comment (BRKalow) [reply src/contexts/__tests__/current-product.test.tsx:23]: Looks great, :shipit: !

**Comment by BRKalow** [comment]
clap

**Comment by ashleemboyer** [comment]
Merging because I'm going heads down to Asana land and don't want to continue blocking any other work ❤️
